iree/turbine/kernel/wave/expansion/expansion_utils.py

```diff
@@ -0,0 +1,253 @@
+
+
+def get_reshape_dim_queries(
+    reshape: Reshape,
+    metadata: ExpansionMetadata,
+    dim_scaling: dict[IndexSymbol, int],
+    nodes_to_expand: list[tuple[CustomOp, dict[IndexSymbol, int]]],
+):
+    """
+    When expanding a reshape, we have to expand the arguments of the reshape and then concatenate them together
+    for the expanded node. Say we have a node with indexing dims = [M, N] with vector shapes m=8, n=2 and
+    the reshape wants to map it to m=4, n=4. So we start by expanding the node
+    node: {m = 0, n = 0}
+        arg: {m = 0, n = 0}
+        arg: {m = 0, n = 1}
+    node: {m = 1, n = 0}
+        arg: {m = 0, n = 0}
+        arg: {m = 0, n = 1}
+    node: {m = 2, n = 0}
+        arg: {m = 1, n = 0}
+        arg: {m = 1, n = 1}
+    node: {m = 3, n = 0}
+        arg: {m = 1, n = 0}
+        arg: {m = 1, n = 1}
+    ...
+    In general,
+    For the (m = i, n = j) expansion of the reshape node, we expand the arguments of the reshape node
+    using the following recipe:
+    - if m_src < m_dst, => we have a one to many mapping from source to destination
+        so we expand the arguments along m = i // (m_dst / m_src) and we expand the argument only once.
+    - if m_src > m_dst, => we have a many to one mapping from source to destination
+        so we expand the arguments along m = i * (m_src / m_dst), ... and we expand the argument m_dst / m_src times.
+
+    In situations where the argument has been expanded along the same dimension, we reuse the expanded node
+    by making use of the context.
+    """
+
+    dim_combinations = {}
+    for dim, value in metadata.dim_query.items():
+        if dim not in reshape.target_vector_shape:
+            continue
+        if reshape.vector_shapes[dim] < reshape.target_vector_shape[dim]:
+            scale_factor = (
+                reshape.target_vector_shape[dim] // reshape.vector_shapes[dim]
+            )
+            dim_combinations[dim] = [value // scale_factor]
+        else:
+            scale_factor = (
+                reshape.vector_shapes[dim] // reshape.target_vector_shape[dim]
+            )
+            begin = value * scale_factor
+            dim_combinations[dim] = list(range(begin, begin + scale_factor))
+    reshape_dim_combinations = list(itertools.product(*dim_combinations.values()))
+    return [
+        {dim: val for dim, val in zip(dim_combinations.keys(), combination)}
+        for combination in reshape_dim_combinations
+    ]
+
+
+def remove_original_nodes(leaf_nodes: list[CustomOp]):
+    """
+    Remove the original nodes from the graph.
+    """
+    stack = leaf_nodes
+    while stack:
+        node = stack.pop(0)
+        inputs, _ = get_inputs(node.fx_node, None)
+        for input in inputs:
+            stack.append(get_custom(input))
+        if not node.users:
+            if node.fx_node._erased:
+                continue
+            node.graph.erase_node(node.fx_node)
```

Comment:
FYI, I tried patching this in and now get errors in a previously functioning kernel where a GetResult node has no value during `remove_original_nodes`. It looks like the node gets revisited even if it's already been erased, so this would happen any time it has multiple users, so I think you want to move the check if the node has already been erased to the beginning of the loop (also, nit: this is a fifo queue, not a stack, and it's got custom ops, not fx nodes)


```suggestion
def remove_original_nodes(leaf_nodes: list[CustomOp]):
    """
    Remove the original nodes from the graph.
    """
    queue = leaf_nodes
    while queue:
        custom = queue.pop(0)
        if custom.fx_node._erased:
            continue
        inputs, _ = get_inputs(custom.fx_node, None)
        for input in inputs:
            queue.append(get_custom(input))
        if not custom.users:
            custom.graph.erase_node(custom.fx_node)
```

Reply:
Thanks! Good catch, I will incorporate these changes.